Refactor target handling & match resolving in `RulesController`

lb- opened the pull request:
This PR is intentionally split out as the final enabler to complete #11045 - setting up a robust `RulesController` that can be used beyond our internal admin use cases.

This refactors some of the existing `RulesController` code, adds unit tests and supports a more explicit `match` mechanism as a response to feedback on previous PRs. Adds two new `match` modes; `one` & `not`.

## Summary of changes

Each change is a discrete commit for easier reviewing.

1. Disable an ESLint rule in test/storybook files as it's not configured well - see https://github.com/wagtail/wagtail/issues/13378 (Update: This has since been fixed)
2. Refactor `RulesController` mechanism for referencing the nearest `form` element, minor improvement that just makes the code a bit easier to follow.
3. Enhance `RulesController` TypeScript & JSDoc to better reference common entities.
4. Refactor `RulesController` effect handling to avoid the large amounts of duplication between show/enable.
5. Refactor `RulesController` match code, this is to address the current awkward API where you had to pass the match value in via the rules with an empty string key, in response to [comment](https://github.com/wagtail/wagtail/pull/12667#discussion_r2269231623). Instead, we will support the match value being declared as the controller value, an action param or on the target element, similar to the rules. Finally, this also includes TWO new match values `not` & `one`, to complete the ability to have these rules match in more flexible ways.

This also includes a large amount of additional unit tests to cover event dispatching/cancelling, various combinations of the new match value approach and additional edge case checks. It also includes a large rework of the Storybook examples with added use cases and a better layout.

## How to test

This PR should not change the existing usage of `RulesController`, you can test these use cases as follows.

1. Within a page, access the privacy modal, check that switching between the privacy modes (e.g. password), should show/hide various fields that match the mode. As per https://github.com/wagtail/wagtail/pull/12667
2. Within a collection, access the privacy modal, check fields as above. As per https://github.com/wagtail/wagtail/pull/12667
3. Within the image URL generator, select various image filter options and check the width/height/closeness fields are disabled/enabled accordingly. As per https://github.com/wagtail/wagtail/pull/11202

lb- committed:
RulesController - Refactor form cache & effect handling

- Add `RuleEntry` type & enhance JSDoc
- Refactor form cache to use a simpler `getter` approach to the form cache & add error handling
- Pull out the effect handling into a shared method to process targets
- Allow for the individual hidden/disabled handling to be declared in isolation
- Allow the matching functions to be added to more easily
- Refactor unit tests for event listeners
- Add unit tests for events being used to cancel effects being applied

laymonage reviewed (commented):
Thank you! I like the extraction of mappings from the effect/match type to their corresponding components, although I'm not sure how to feel about `processTarget()`. See my comment for more details.

lb- committed:
RulesController - Add support for 'one' & 'not' match

- Add unit test coverage
- Refactor storybook examples to include more robust examples & one for 'not'

lb- commented on client/src/controllers/RulesController.ts:
```diff
@@ -122,49 +147,71 @@ export class RulesController extends Controller<
+    const { match, rules } = this.parseRules(target, effect);
+    const result = matchers[match](rules);
+
+    const apply = effectHandler(target, result);
```
Note: These recommendations have been mostly adopted. Closing thread.

lb- commented:
I have rebased, cleaned up the changes.
The frontend CI is flagging warnings against `jsx-a11y/label-has-associated-control` but this rule is ignored in storybook files so there might be some caching issue.

lb- commented:
@laymonage  - if possible, this is ready for another review based on feedback (applied) from late last year.

laymonage commented on client/src/controllers/RulesController.ts:
```diff
@@ -72,87 +90,161 @@ export class RulesController extends Controller<
+      const form = 'form' in element ? element.form : element.closest('form');
+      if (form) {
+        this.formCache = form;
+      }
```
This is redundant since we'll check `if (this.formCache)` below anyway

```suggestion
      this.formCache = form
```

laymonage committed:
Update eslint-config-wagtail

laymonage commented on client/src/controllers/RulesController.ts:
```diff
@@ -72,87 +111,163 @@ export class RulesController extends Controller<
     const checkFn = ([fieldName, allowedValues]) => {
       // Forms can have multiple values for the same field name
-      const values = formData.getAll(fieldName);
+      const values = new FormData(this.form).getAll(fieldName);
```
I think we can construct the `FormData` outside the `checkFn` so that we don't create it on every loop. This is similar to the existing behaviour.

```suggestion
    const formData = new FormData(this.form);
    const checkFn = ([fieldName, allowedValues]) => {
      // Forms can have multiple values for the same field name
      const values = formData.getAll(fieldName);
```

laymonage reviewed (commented):
Just some minor tweaks to suggest, otherwise this looks good to me, thanks!

laymonage committed:
Release note for #13379

lb- commented:
Thanks @laymonage - good feedback above, thanks for the final sense check and merge.
